### PR TITLE
8257197: Add additional verification code to PhaseCCP

### PR DESCRIPTION
**Pull request by eme64**
We have had many bugs that could be tracked back to missing optimizations during PhaseCCP.
Often the problem is that a node `x` has an input (of input of input) `y` which is modified, but `y` does not notify `x` (does not push it to the worklist). For one this is a missed opportunity to optimize, but it can also lead to failed assumptions later: often we assume that all that can be optimize is already optimized.
This verification helps us debug faster, and can also help when `Value` optimizations suddenly do further-reaching traversals, which would require further-reaching notifications.

Sadly, the verification is not total: we have some exceptions. Especially for `LoadNode`, which perform a walk up the memory inputs, which can go arbitrarily far, to look for relates `StoreNode`s. Notification would thus have to put very many nodes on the worklist. The question is if the potential additional optimization is worth the compile-time. If we decided yes, then one we might want to implement a listener-style notification: when a node visits inputs during `Value`, it could subscribe to all (or the relevant) visited input-nodes for future updates. Currently, we just mostly do fixed 1-hop or 2-hop notification of output nodes.

FYI: I plan to do a similar verification, and a refactoring of `PhaseCCP::push_child_nodes_to_worklist` and `PhaseIterGVN::add_users_to_worklist` in [JDK-8298094](https://bugs.openjdk.org/browse/JDK-8298094).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8257197](https://bugs.openjdk.org/browse/JDK-8257197): Add additional verification code to PhaseCCP

### Reviewers
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [db08362b](https://git.openjdk.org/jdk/pull/11529/files/db08362bdb4ab82d458164fd934dbaea531c5b6c)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11529/head:pull/11529` \
`$ git checkout pull/11529`

Update a local copy of the PR: \
`$ git checkout pull/11529` \
`$ git pull https://git.openjdk.org/jdk pull/11529/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11529`

View PR using the GUI difftool: \
`$ git pr show -t 11529`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11529.diff">https://git.openjdk.org/jdk/pull/11529.diff</a>

</details>

### CODE REVIEW

**Review by chhagedorn — APPROVED** [review]
This is a nice additional verification! I've tried your patch out and it would have indeed found https://github.com/openjdk/jdk19/pull/65 and https://github.com/openjdk/jdk/pull/11448 (in both bugs we've missed to re-add nodes to the CCP worklist).

It's a simple but powerful verification pass, so I think it is okay to add some exceptions in order to make this as stable as possible to avoid false positives/okay-to-miss optimizations.

File: src/hotspot/share/opto/phaseX.cpp [hunk 1]
```diff
@@ -1798,8 +1800,56 @@ void PhaseCCP::analyze() {
...
+        // MemNode::can_see_stored_value looks up through many memory nodes,
+        // which means we would need to notify modifications from far up in
+        // the inputs all the way down to the LoadNode. We don't do that.
+        continue;
+      }
+      tty->print_cr("Missed optimization (PhaseCCP):");
+      n->dump_bfs(1, 0, "");
+      tty->print_cr("Current type:");
+      told->dump_on(tty);
+      tty->print_cr("");
```
[the remark this answers is not in the document]
Comment (eme64) [reply src/hotspot/share/opto/phaseX.cpp]: :+1:

File: src/hotspot/share/opto/phaseX.cpp [hunk 2]
```diff
@@ -1798,8 +1800,56 @@ void PhaseCCP::analyze() {
...
+      if (n->is_Load()) {
+        // MemNode::can_see_stored_value looks up through many memory nodes,
+        // which means we would need to notify modifications from far up in
+        // the inputs all the way down to the LoadNode. We don't do that.
+        continue;
+      }
+      tty->print_cr("Missed optimization (PhaseCCP):");
+      n->dump_bfs(1, 0, "");
+      tty->print_cr("Current type:");
+      told->dump_on(tty);
```
[the remark this answers is not in the document]
Comment (eme64) [reply src/hotspot/share/opto/phaseX.cpp:1835]: will leave it as is. extra new-line should already help with readability

File: src/hotspot/share/opto/phaseX.cpp [hunk 3]
```diff
@@ -1798,8 +1800,56 @@ void PhaseCCP::analyze() {
...
+      tty->print_cr("");
+      tty->print_cr("Optimized type:");
+      tnew->dump_on(tty);
+      tty->print_cr("");
```
[the remark this answers is not in the document]
Comment (eme64) [reply src/hotspot/share/opto/phaseX.cpp]: :+1:

File: src/hotspot/share/opto/phaseX.cpp [hunk 4]
```diff
@@ -1798,8 +1800,56 @@ void PhaseCCP::analyze() {
...
+      failure = true;
+    }
+  }
+  // If you get this assert, check if the node was notified of changes in
+  // the inputs. See PhaseCCP::push_child_nodes_to_worklist
```
[the remark this answers is not in the document]
Comment (eme64) [reply src/hotspot/share/opto/phaseX.cpp]: Added your text, thanks

File: src/hotspot/share/opto/phaseX.cpp [hunk 5]
```diff
@@ -1798,8 +1800,56 @@ void PhaseCCP::analyze() {
...
+        if (t0->_lo == t1->_lo && t0->_hi == t1->_hi) {
+          continue; // ignore int widen
+        }
+      } else if (told->isa_long() && tnew->isa_long() ) {
+        const TypeLong *t0 = told->is_long();
+        const TypeLong *t1 = tnew->is_long();
+        if (t0->_lo == t1->_lo && t0->_hi == t1->_hi) {
+          continue; // ignore long widen
+        }
+      }
```
[the remark this answers is not in the document]
Comment (eme64) [reply src/hotspot/share/opto/phaseX.cpp]: refactored it

**Review by chhagedorn — APPROVED** [review]
Thanks for doing the updates, looks good!

**Review by chhagedorn — APPROVED** [review]

**Review by TobiHartmann — APPROVED** [review]
Looks good!

**Comment by eme64** [comment]
Thanks @TobiHartmann @chhagedorn @vnkozlov for the reviews!
/integrate

**Pull request closed without merging**